packages/overture-schema-core/src/overture/schema/core/scoping/__init__.py

```diff
@@ -0,0 +1,394 @@
+Specifically, the value `Side.LEFT` is on the left of a person who is facing forward (toward the
+end of the geometry), and `Side.RIGHT` is likewise on this person's right.
+
+
+Vehicle:
+--------
+Vehicle scoping allows a value to be tied to one or more properties of a vehicle, such as height,
+weight, or number of axles. This enables a wide variety of use cases such as restricting the
+allowed weight of trucks on bridges, limiting the allowed height of vehicles crossing under
+bridges or entering garages, applying differential speed limits to different classes of vehicle,
+*etc.* When a model is decorated with vehicle scoping, a `when.vehicle` field is automatically
+added to the model.
+
+>>> from overture.schema.core.unit import LengthUnit
+>>> from overture.schema.system.primitive import float32
+...
+>>> @scoped(Scope.VEHICLE)
+... class Fare(BaseModel):
+...     value: float32
+...
+>>> fare_schedule: list[Fare] = [
+...     Fare(
+...         value=10,
+...         when=Fare.When(vehicle=[
+...             VehicleAxleCountSelector(
+...                 dimension=VehicleDimension.AXLE_COUNT,
+...                 comparison=VehicleRelation.LESS_THAN,
+...                 value=3
+...             ),
+...             VehicleLengthSelector(
+...                dimension=VehicleDimension.LENGTH,
+...                comparison=VehicleRelation.LESS_THAN_EQUAL,
+...                value=18,
+...                unit=LengthUnit.FT,
+...             )
+...         ]),
+...     ),
+...     Fare(value=30),
+... ]
+
+
+
+Mixing scopes
+==============
+The `@scoped` decorator can mix any desired combination of scopes onto your Pydantic model.
+
+For example, suppose you are modeling a value type that can apply at certain times, along a certain
+sub-segment of a linear path, depending on whether one is traveling forward or backward along the
+path, or any combination of these three. This can be achieved easily with:
+
+>>> @scoped(Scope.TEMPORAL, Scope.GEOMETRIC_RANGE, Scope.HEADING)
+... class MyModel(BaseModel):
+...     pass
+
+
+Optional and required scopes
+============================
+When using the `@scoped` decorator, scopes are optional by default but some or all scopes may be
+made required.
+
+The following example makes all the scopes required:
+
+>>> from enum import Enum
+...
+>>> class SignalType(str, Enum):
+...    STOP_SIGN = 'stop_sign'
+...
+>>> @scoped(required=(Scope.GEOMETRIC_POSITION, Scope.HEADING))
+... class TrafficSignal(BaseModel):
+...    signal_type: SignalType
+
+The following example mixes an optional scope (temporal) with two required scopes (geometric
+position and heading).
+
+>>> @scoped(Scope.TEMPORAL, required=(Scope.GEOMETRIC_POSITION, Scope.HEADING))
+... class TrafficSignal(BaseModel):
+...    signal_type: SignalType
+
+
+The `when` clause
+=================
+For historical reasons, some scope fields are added directly to the decorated model, while others
+are added as children of a synthetic `when` field.
```

Comment:
How should we resolve this inconsistency?

Is the ideal state moving `at`, `between`, and `side` under `when`?

Would it be worth allowing authors to model using the ideal state while introducing a backward-compatibility shim that allows us to migrate our distributed data independently of modeling? (kinda similar to how we pivot JSON Schema to match GeoJSON, but probably at a higher level than JSON Schema generation)

Reply:
I think the ideal state is probably getting rid of `when`.

At the moment it's hard for me to imagine a shim that would quite get the job done at all the required levels. I'm tempted to punt and hope for a beautiful tomorrow where we just get rid of `when` with an MV bump.